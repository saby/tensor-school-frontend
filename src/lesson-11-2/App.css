.App {
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
}
.App__buttons {
    width: 127px;
    display: flex;
    flex-direction: column;
}
.App__image {
    background-image: url(./want.jpg);
    width: 320px;
    height: 180px;
    background-size: cover;
    background-repeat: no-repeat;
}
.App__list {
    flex-basis: 400px;
}
.App__list-big {
    font-size: 2rem;
}
.App__list-small {
    font-size: 1rem;
}
