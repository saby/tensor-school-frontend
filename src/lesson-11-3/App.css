.App {
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
    flex-direction: column;
}

.App__counter {
    position: relative;
    margin-top: 400px;
}

.App__counter::before {
    background-image: url("./count.jpg");
    background-repeat: no-repeat;
    content: '';
    width: 509px;
    height: 457px;
    position: absolute;
    z-index: -1;
    left: -176px;
    bottom: 15px;
}

.App__reset {
    margin-left: 16px;
    margin-bottom: 8px;
}
/* Chrome, Safari, Edge, Opera */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

/* Firefox */
input[type=number] {
    -moz-appearance: textfield;
}
