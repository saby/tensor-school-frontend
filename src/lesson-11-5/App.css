.App {
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
    flex-direction: column;
}
.App__keyboard {
    display: grid;
    grid-template-columns: repeat(4, 40px);
    grid-template-rows: repeat(4, 40px);
    grid-gap: 8px;
}
.App__keyboard > div {
    background-color: rgba(153, 153, 153, 0.2);
    border-radius: 8px;
    line-height: 40px;
    cursor: pointer;
}
.App__keyboard > div:hover {
    box-shadow: 0 0 4px darkgray;
}
.App__keyboard > div:active {
    background-color: rgba(153, 153, 153, 0.4);
}
.App__string {
    margin-bottom: 8px;
    width: 100%;
    box-sizing: border-box;
}
.App__result {
    background: linear-gradient(to right, red, orange, yellow, green, cyan, blue, violet);
    width: 100%;
    height: 40px;
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    font-size: 30px;
    line-height: 40px;
    text-align: right;
}
.App__calculator {
    position: relative;
}
.App__calculator::before {
    position: absolute;
    content: '';
    width: 514px;
    height: 457px;
    top: -50px;
    left: 100%;
    background-image: url("./calc.jpg");
    background-size: cover;
    z-index: -1;
}
