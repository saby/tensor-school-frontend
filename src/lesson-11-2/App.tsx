import './App.css'
import {ReactFragment, useState} from 'react';

type ListType = 'unordered' | 'ordered';
function toggleListType(listType: ListType): ListType {
    return listType === 'unordered' ? 'ordered' : 'unordered';
}

type Color = 'red' | 'violet';
function toggleColor(color: Color): Color {
    return color === 'red' ? 'violet' : 'red';
}

type Font = 'big' | 'small';
function toggleFont(font: Font): Font {
    return font === 'big' ? 'small' : 'big';
}

// TODO убрать
function ListItems(): ReactFragment {
    return <>
        <li>Посмотреть вокруг</li>
        <li>Оглянуться назад</li>
        <li>Рыбов купить</li>
        <li>"Красивое" сказать</li>
    </>;
}

function App() {
    const [listType, setListType] = useState<ListType>('unordered');
    const [color, setColor] = useState<Color>('red');
    const [font, setFont] = useState<Font>('big');
    return (
        <div className="App">
            {/* Изменения делать только ниже */}
            <div className={"App__list App__list-" + font} data-testid="list"
                style={{
                    color: color
                }}>
                {listType === 'unordered' ?
                    <ul>{ListItems()}</ul> :
                    <ol>{ListItems()}</ol>}
            </div>
            {/* Изменения делать только выше */}
            <div className="App__buttons" data-testid="buttons">
                <button data-testid="listType"
                        onClick={() => setListType(toggleListType)}>
                    {'Хочу ' + (listType === 'unordered' ? 'цифры' : 'точки')}
                </button>
                <button data-testid="color"
                        onClick={() => setColor(toggleColor)}>
                    {'Хочу ' + (color === 'red' ? 'фиолетовый' : 'красный' )}
                </button>
                <button data-testid="fontSize"
                        onClick={() => setFont(toggleFont)}>
                    {'Хочу ' + (font === 'big' ? 'маленькие' : 'большие' )}
                </button>
            </div>
            <div className="App__image"/>
        </div>
    );
}

export default App;
