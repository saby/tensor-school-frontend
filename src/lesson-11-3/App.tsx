import './App.css';
import {useState, useEffect, ChangeEvent} from 'react';

// TODO оставить только верстку с testid
// Убрать зависимости react и state
function App() {
    const [counter, setCounter] = useState(0);
    const [delay, setDelay] = useState(1000);

    useEffect(() => {
        const id = setInterval(() => {
            setCounter((count) => count + 1);
        }, delay);
        return () => clearInterval(id);
    }, [delay])

    return (
        <div className="App">
            <div className="App__counter">
                Counter: {counter}
                <button data-testid="reset"
                        className="App__reset"
                        onClick={() => setCounter(0)}>
                    Reset
                </button>
            </div>
            <div className="App__input">
                <input data-testid="input"
                       type="number"
                       pattern="\d"
                       onChange={(e: ChangeEvent<HTMLInputElement>) => setDelay(Number(e.target.value))}
                       value={delay}/>
            </div>
        </div>
    );
}

export default App;
