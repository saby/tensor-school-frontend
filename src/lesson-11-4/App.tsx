import Table from './Table'
import {DATA, IData} from './Data';
import './App.css'
import {useCallback, useRef, useState} from 'react';
import SortContext, {ISortContext, ISortFields, Sort} from './SortContext';

const SORT_MULTIPLIER: Record<Sort, number> = {
    'ASC': 1,
    'DESC': -1,
    '': 0
}

const EMPTY_SORT: ISortFields = {
    date: '',
    name: '',
    cost: ''
}

// TODO убрать
function filterData(search: string) {
    return [...DATA].filter(({name }) => name.toLowerCase().search(search.toLowerCase()) !== -1);
}

function App() {
    const [search, setSearch] = useState('')
    const [data, setData] = useState(DATA);
    const searchRef = useRef(search);
    const [sort, setSort] = useState<ISortContext>({
        ...EMPTY_SORT,
        changeSort(name: string, sortValue: Sort): void {
            const newSort = {...EMPTY_SORT, changeSort: sort.changeSort, [name]: sortValue}
            // TODO убрать
            setSort(newSort)
            setData(sortData(filterData(searchRef.current), newSort));
            // TODO END убрать
        }
    })


    const sortData = useCallback((data: IData[], sort: ISortContext) => {
        // TODO убрать
        return data.sort((a, b) => {
            const compare = (field: keyof ISortFields) => {
                if (sort[field]) {
                    return a[field] < b[field] ? -SORT_MULTIPLIER[sort[field]] : SORT_MULTIPLIER[sort[field]];
                } else {
                    return 0;
                }
            }
            return compare('date') + compare('name') + compare('cost')
        });
        // TODO END убрать
    }, []);

    const onSearch = useCallback((searchString: string) => {
        // TODO убрать
        setData(sortData(filterData(searchString), sort));
        setSearch(searchString);
        // TODO END убрать
        searchRef.current = searchString;

    }, [sort, sortData])

    return (
        <div className="App">
            <SortContext.Provider value={sort}>
                {/* TODO убрать атрибуты */}
                <Table data={data} searchValue={search} onSearch={onSearch}/>
            </SortContext.Provider>
        </div>
    );
}

export default App;
