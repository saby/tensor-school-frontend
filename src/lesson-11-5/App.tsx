import './App.css';
import {useState, ChangeEventHandler, useCallback, MouseEventHandler, KeyboardEventHandler} from 'react';
// TODO наверно убрать все кроме верстки с className и testid
const onlyExpressionSymbols = /[^0-9+\-=*/.]/g;
const doubleSigns = /([-+=*/.])[+=*/.]*/g;
const doubleSub = /(-)-*/g;
const startSign = /^[+=*/]/;
const SIGNS: Record<string, string> = {
    '÷': '/',
    '×': '*',
    '−': '-'
}

function App() {
    const [expression, setExpression] = useState('');
    const [result, setResult] = useState('');

    const handleExpression = useCallback((expression) => {
        try {
            // потому шо лень
            // eslint-disable-next-line no-eval
            const result = eval(expression);
            setResult(+result.toFixed(10) + '');
        } catch {}
    }, [])

    const updateExpression = useCallback((value: string) => {
        const newValue = value.replace(startSign, '')
            .replace(onlyExpressionSymbols, '')
            .replace(doubleSigns, '$1')
            .replace(doubleSub, '$1');
        if (newValue !== expression) {
            if (newValue[newValue.length - 1] === '=') {
                handleExpression(newValue.slice(0, -1));
            } else {
                setExpression(newValue);
            }
        }
    }, [expression, setExpression, handleExpression])

    const changeString: ChangeEventHandler<HTMLInputElement> = useCallback((e) => {
        const {value} = e.target;
        updateExpression(value);
    }, [updateExpression])

    const keyboardClick = useCallback<MouseEventHandler<HTMLDivElement>>((event) => {
        const {target} = event;
        const HTMLTarget = target as HTMLElement;
        if (HTMLTarget.children.length === 0) {
            const addition = SIGNS[HTMLTarget.textContent as string] || HTMLTarget.textContent;
            updateExpression(expression + addition);
        }
    }, [expression, updateExpression]);

    const onSubmit = useCallback<KeyboardEventHandler>(
        (event) => event.key === "Enter" && handleExpression(expression),
        [handleExpression, expression]
    );

    return (<div className="App">
        <div className="App__calculator">
            <div data-testid="result"
                 className="App__result">
                {result}
            </div>
            <input data-testid="input"
                   className="App__string"
                   value={expression}
                   onChange={changeString}
                   onKeyPress={onSubmit}/>
            <div
                data-testid="keyboard"
                className="App__keyboard"
                onClick={keyboardClick}>
                <div>7</div><div>8</div><div>9</div><div>÷</div>
                <div>4</div><div>5</div><div>6</div><div>×</div>
                <div>1</div><div>2</div><div>3</div><div>−</div>
                <div>0</div><div>.</div><div>=</div><div>+</div>
            </div>
        </div>
    </div>)
}

export default App;
