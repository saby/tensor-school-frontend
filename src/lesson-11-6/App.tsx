import {UserContext, IUserContext, IUserData} from "./UserContext";
import {useState} from "react";
import User from "./User";
import './App.css'

function App() {
    const [userData, setData] = useState<IUserContext>(() => ({
        about: "человек",
        birthdate: new Date(1997, 5, 13),
        books: "дюна",
        email: "myemail@mail.com",
        films: "дюна",
        firstName: "Иванов",
        games: "Ведьмак",
        gender: 'мужской',
        hobby: "реверс-инжиниринг",
        lastName: "Иван",
        location: "г. Ярославль",
        music: "mood music",
        patronym: "Иванович",
        phone: "+701234567890",
        school: {
            name: '322',
            address: 'Ярославль, ул. Пушкина д. Колотушкина',
            startDate: new Date(2004, 8),
            finishDate: new Date(2015, 4)
        },
        university: {
            name: 'ЯрГУ им.П.Г.Демидова',
            address: 'Ярославль, ул. Союзная 144',
            startDate: new Date(2015, 8),
            finishDate: new Date(2019, 7, 31),
            specialization: 'Компьютерные науки'
        },
        work: "Тензор",
        updateUser(updateUserData: Partial<IUserData>): void {
            setData((userData) => ({...userData, ...updateUserData}))
        }
    }))

    return (
        <div className="App">
            <UserContext.Provider value={userData}>
                <User/>
            </UserContext.Provider>
        </div>
    );
}

export default App;
