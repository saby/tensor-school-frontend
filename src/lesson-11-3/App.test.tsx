import App from './App';
import {act, fireEvent, render} from '@testing-library/react';

afterEach(() => {
    jest.useRealTimers();
})
describe('Render', () => {
    it('Проверка верстки по умолчанию', () => {
        const {container} = render(<App/>);

        expect(container).toMatchSnapshot();
    })

    it('Проверка стандартной задержки в 1 сек', () => {
        jest.useFakeTimers();
        const {container} = render(<App/>);

        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);

        act(() => {
            jest.advanceTimersByTime(1000);
        })

        expect(container).toMatchSnapshot();
    })


    it('Проверка изменения задержки', () => {
        jest.useFakeTimers();
        const {container, getByTestId} = render(<App/>);

        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);

        const input = getByTestId('input');
        fireEvent.change(input, {target: {value: '500'}})

        expect(clearInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenCalledTimes(2);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 500);

        act(() => {
            jest.advanceTimersByTime(500);
        })
        expect(container).toMatchSnapshot();
    })

    it('Проверка сброса счетчика', () => {
        jest.useFakeTimers();
        const {container, getByTestId} = render(<App/>);

        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);

        act(() => {
            jest.advanceTimersByTime(1000);
        })

        expect(container).toMatchSnapshot();

        const reset = getByTestId('reset');
        fireEvent.click(reset);

        expect(container).toMatchSnapshot();
    })
})
