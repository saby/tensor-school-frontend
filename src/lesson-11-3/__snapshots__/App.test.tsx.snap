// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Проверка верстки по умолчанию 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App__counter"
    >
      Counter: 
      0
      <button
        class="App__reset"
        data-testid="reset"
      >
        Reset
      </button>
    </div>
    <div
      class="App__input"
    >
      <input
        data-testid="input"
        pattern="\\\\d"
        type="number"
        value="1000"
      />
    </div>
  </div>
</div>
`;

exports[`Render Проверка изменения задержки 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App__counter"
    >
      Counter: 
      1
      <button
        class="App__reset"
        data-testid="reset"
      >
        Reset
      </button>
    </div>
    <div
      class="App__input"
    >
      <input
        data-testid="input"
        pattern="\\\\d"
        type="number"
        value="500"
      />
    </div>
  </div>
</div>
`;

exports[`Render Проверка сброса счетчика 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App__counter"
    >
      Counter: 
      1
      <button
        class="App__reset"
        data-testid="reset"
      >
        Reset
      </button>
    </div>
    <div
      class="App__input"
    >
      <input
        data-testid="input"
        pattern="\\\\d"
        type="number"
        value="1000"
      />
    </div>
  </div>
</div>
`;

exports[`Render Проверка сброса счетчика 2`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App__counter"
    >
      Counter: 
      0
      <button
        class="App__reset"
        data-testid="reset"
      >
        Reset
      </button>
    </div>
    <div
      class="App__input"
    >
      <input
        data-testid="input"
        pattern="\\\\d"
        type="number"
        value="1000"
      />
    </div>
  </div>
</div>
`;

exports[`Render Проверка стандартной задержки в 1 сек 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App__counter"
    >
      Counter: 
      1
      <button
        class="App__reset"
        data-testid="reset"
      >
        Reset
      </button>
    </div>
    <div
      class="App__input"
    >
      <input
        data-testid="input"
        pattern="\\\\d"
        type="number"
        value="1000"
      />
    </div>
  </div>
</div>
`;
